File: server/scripts/generate-snapshots.ts
#!/usr/bin/env tsx
/** Regenerate API response snapshots for regression testing (Node.js platform). */

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";
import request from "supertest";
import { initDatabase } from "../src/db/index.js";
import { createApp } from "../src/app.js";
import { normalizeSnapshot } from "../src/utils/snapshot-normalize.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const SNAPSHOT_DIR = path.resolve(__dirname, "../../tests/snapshots");

initDatabase();
const app = createApp();

async function writeSnapshot(name: string, data: unknown) {
  const normalized = normalizeSnapshot(data);
  const out = path.join(SNAPSHOT_DIR, `${name}.json`);
  fs.writeFileSync(out, `${JSON.stringify(normalized, null, 2)}\n`, "utf-8");
  console.log(`Wrote ${out}`);
}

async function main() {
  fs.mkdirSync(SNAPSHOT_DIR, { recursive: true });

  const root = await request(app).get("/api");
  await writeSnapshot("root", root.body);

  const health = await request(app).get("/api/v1/health");
  await writeSnapshot("health", health.body);

  const integrations = await request(app).get("/api/v1/integrations/status");
  await writeSnapshot("integrations_status", integrations.body);

  const demoCredit = await request(app).get("/api/v1/assess/demo").query({ audience: "credit_team" });
  await writeSnapshot("demo_assessment_credit", demoCredit.body);

  const demoRisk = await request(app).get("/api/v1/assess/demo").query({ audience: "risk_team" });
  await writeSnapshot("demo_assessment_risk", demoRisk.body);

  const policies = await request(app).get("/api/v1/policies/catalog").query({ sector: "auto_components" });
  await writeSnapshot("policies_auto", policies.body);

  const bureau = await request(app).post("/api/v1/integrations/bureau/pull").query({
    gstin: "27AABCS1234F1Z5",
    pan: "AABCS1234F",
    business_name: "Shree Ganesh Auto Components Pvt Ltd",
  });
  await writeSnapshot("bureau_pull", bureau.body);

  const tax = await request(app).post("/api/v1/integrations/tax/verify").query({
    gstin: "27AABCS1234F1Z5",
    pan: "AABCS1234F",
  });
  await writeSnapshot("tax_verify", tax.body);

  const architecture = await request(app).get("/api/v1/agents/architecture");
  await writeSnapshot("agents_architecture", architecture.body);

  const credentials = await request(app).get("/api/v1/auth/demo-credentials");
  await writeSnapshot("demo_credentials", credentials.body);

  const login = await request(app)
    .post("/api/v1/auth/login")
    .send({ email: "rajesh@shreeganesh.in", password: "MSME@2026" });
  const assess = await request(app)
    .post("/api/v1/msme/assess/quick")
    .set("Authorization", `Bearer ${login.body.access_token}`);
  await writeSnapshot("msme_orchestration", assess.body.agent_insights);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scripts/generate_snapshots.py
#!/usr/bin/env python3
"""Regenerate API response snapshots for regression testing."""

from __future__ import annotations

import asyncio
import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

from httpx import ASGITransport, AsyncClient

from app.main import app

SNAPSHOT_DIR = ROOT / "tests" / "snapshots"
VOLATILE_KEYS = {"assessment_id", "generated_at"}


def normalize(obj, parent_key: str | None = None):
    if isinstance(obj, dict):
        return {
            k: ("<UUID>" if k in VOLATILE_KEYS else normalize(v, k))
            for k, v in obj.items()
        }
    if isinstance(obj, list):
        return [normalize(item, parent_key) for item in obj]
    if isinstance(obj, str) and parent_key in VOLATILE_KEYS:
        return "<TIMESTAMP>" if "T" in obj and re.match(r"\d{4}-\d{2}", obj) else obj
    return obj


async def fetch_snapshot(client: AsyncClient, name: str, method: str, path: str, **kwargs) -> None:
    if method == "GET":
        response = await client.get(path, **kwargs)
    else:
        response = await client.post(path, **kwargs)
    data = normalize(response.json())
    out = SNAPSHOT_DIR / f"{name}.json"
    out.write_text(json.dumps(data, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    print(f"Wrote {out}")


async def main() -> None:
    SNAPSHOT_DIR.mkdir(parents=True, exist_ok=True)
    transport = ASGITransport(app=app)
    async with AsyncClient(transport=transport, base_url="http://test") as client:
        await fetch_snapshot(client, "root", "GET", "/")
        await fetch_snapshot(client, "health", "GET", "/api/v1/health")
        await fetch_snapshot(client, "integrations_status", "GET", "/api/v1/integrations/status")
        await fetch_snapshot(client, "demo_assessment_credit", "GET", "/api/v1/assess/demo", params={"audience": "credit_team"})
        await fetch_snapshot(client, "demo_assessment_risk", "GET", "/api/v1/assess/demo", params={"audience": "risk_team"})
        await fetch_snapshot(client, "policies_auto", "GET", "/api/v1/policies/catalog", params={"sector": "auto_components"})
        await fetch_snapshot(client, "bureau_pull", "POST", "/api/v1/integrations/bureau/pull", params={
            "gstin": "27AABCS1234F1Z5", "pan": "AABCS1234F", "business_name": "Shree Ganesh Auto Components Pvt Ltd",
        })
        await fetch_snapshot(client, "tax_verify", "POST", "/api/v1/integrations/tax/verify", params={
            "gstin": "27AABCS1234F1Z5", "pan": "AABCS1234F",
        })


if __name__ == "__main__":
    asyncio.run(main())
